envconfig.go

```diff
@@ -32,75 +54,227 @@ type Decoder interface {
 	Decode(value string) error
 }
 
+// ProcessFunc vistor function definition while traversing the list of options
+type ProcessFunc func(field reflect.Value, tof reflect.StructField, fieldName string, key string, alt_key string, def string, required string) error
+
 func (e *ParseError) Error() string {
 	return fmt.Sprintf("envconfig.Process: assigning %[1]s to %[2]s: converting '%[3]s' to type %[4]s", e.KeyName, e.FieldName, e.Value, e.TypeName)
 }
 
-// Process populates the specified struct based on environment variables
-func Process(prefix string, spec interface{}) error {
+// Visit vists all the configuration options, calling the process function on each
+func Visit(prefix string, spec interface{}, process ProcessFunc) error {
```

Comment:
While I agree the main mission of envconfig is to provide parsing of config values from the environment, I don't see enabling individuals to extend that capability as a big departure from that mission. The visitor pattern is only for the config structure and just allows individual project a mechanism (for a small percent of people I would guess) to add capability if they so desire.

For example, as I stated elsewhere, I am developing an output current config settings visitor function. This may not be something that envconfig wants as part of its core code. But without exposing the Visitor func and the DocumentInfo struct I end up either having a private fork of envconfig or have to attempt to get this merged into envconfig.

Exposing these items seems a small price and a small mission growth for this benefit.